src/lib/zkapp.ts

```diff
@@ -93,84 +110,277 @@ function method<T extends SmartContract>(
         () => {
-          // inside prover / compile, the method is always called with the public input as first argument
-          // -- so we can add assertions about it
-          let publicInput = actualArgs[0];
-          actualArgs = actualArgs.slice(1);
-
-          // outside a transaction, just call the method, but check precondition invariants
-          let result = method.apply(this, actualArgs);
-          checkPublicInput(publicInput, this.self);
-
-          // check the self party right after calling the method
-          // TODO: this needs to be done in a unified way for all parties that are created
-          assertPreconditionInvariants(this.self);
-          cleanPreconditionsCache(this.self);
-          assertStatePrecondition(this);
-          return result;
+          if (inCheckedComputation()) {
+            // important to run this with a fresh party everytime, otherwise compile messes up our circuits
+            // because it runs this multiple times
+            let [, result] = Mina.currentTransaction.runWith(
+              {
+                sender: undefined,
+                parties: [],
+                fetchMode: inProver() ? 'cached' : 'test',
+                isFinalRunOutsideCircuit: false,
+              },
+              () => {
+                // inside prover / compile, the method is always called with the public input as first argument
+                // -- so we can add assertions about it
+                let publicInput = actualArgs[0];
+                actualArgs = actualArgs.slice(1);
+                let party = this.self;
+
+                // the blinding value is important because otherwise, putting callData on the transaction would leak information about the private inputs
+                let blindingValue = Circuit.witness(Field, getBlindingValue);
+                // it's also good if we prove that we use the same blinding value across the method
+                // that's why we pass the variable_ (not the constant) into a new context
+                let context = memoizationContext() ?? {
+                  memoized: [],
+                  currentIndex: 0,
+                };
+                let [, result] = memoizationContext.runWith(
+                  { ...context, blindingValue },
+                  () => method.apply(this, actualArgs)
+                );
+
+                // connects our input + result with callData, so this method can be called
+                let callDataFields = methodArgumentsToFields(
+                  methodIntf,
+                  actualArgs
+                );
+                if (methodIntf.returnType) {
+                  callDataFields.push(
+                    ...methodIntf.returnType.toFields(result)
+                  );
+                }
+                callDataFields.push(getMethodId(methodIntf, methodIndex));
+                callDataFields.push(blindingValue);
+                party.body.callData = Poseidon.hash(callDataFields);
```

Comment:
This whole code block is the case that we're in a circuit. So, these commands are what is added to the circuit on top of the user code in a `@method`. It is also how a called method is executed inside its own prover (but not in the prover of the caller -- that case is further below!). On this line, we're populating our own `callData` field -- as explained [in the RFC](https://github.com/o1-labs/snarkyjs/issues/303), this is done every time, to make the method callable by default